### DISCUSSION TITLE
useFetch -> How to reuse args and return type ?

### DISCUSSION THREAD

**Discussion by Zebnastien** [nuxt/nuxt#26464] [Questions]
Hi there,

I am trying to build a composable with embedded useFetch and other stuff.

```ts
export const useBackend: typeof useFetch = async (url, options?, params?) => {
    const { data, error } = await useFetch(url, {
        headers: useRequestHeaders(["cookie"]),
        ...options,
    });

    const items = useState(() => data);

    // Other stuff using params that will might be returned...

    return { data, error, items, otherStuff };
};
```

I would like to keep route autocompletion from useFetch args like this:

![image](https://github.com/nuxt/nuxt/assets/64729974/b4efd4d9-dd2d-4d89-b0ac-9dc478d82578)

...with my own function:
```ts
{ data } = await useBackend(url)
```

Also I would like to keep the dynamic typing on useFecth return like this: 

![image](https://github.com/nuxt/nuxt/assets/64729974/cfd09de3-1afd-482d-89f7-695647f1bd70)

I have tried many ways but involving Extracted types and Parameters<typeof...> but I am not very skilled in TypeScript.
Is there a way to do my own composable using args and return types inspired from useFetch ?

**Comment by Zebnastien** [comment]
I finally found a simple solution. Just import { NitroFetchRequest } from "nitropack":

```ts
import type { NitroFetchRequest } from "nitropack";

export default async function <U extends NitroFetchRequest>(url: U) {
    const response = await useFetch(url);

    return response;
}
```

**Comment by Zebnastien** [comment]
Ok, now I am searching how to do the same for useFetchOptions:
I want to keep options autocompletion.

Something like this (not working):

```ts
export default async function <
    U extends NitroFetchRequest,
    O extends UseFetchOptions
>(url: U, options: O) {
    const response = await useFetch(url, { ...options } as UseFetchOptions);
    return response;
}
```

**Comment by Zebnastien** [comment]
Ok, I ended with this:

```ts
import type { NitroFetchRequest } from "nitropack";

type useFetchParams = Parameters<typeof useFetch>;

export default async function <
    U extends NitroFetchRequest,
    O extends useFetchParams[1]
>(url: U, options?: O) {
    const response = await useFetch(url, { ...options });

    return response;
}
```

But I still got a TS error when spreading options...

**Comment by xjccc** [comment]
```
const response = await useFetch(url as string, { ...options })
```
